### PR TITLE
Fix for #2620. Added Nullable type handling for parameters.

### LINKED ISSUE #2620: Conversion Exception
I'm facing an exception so I added a test to reproduce it to this file: [ValueConversionTests.cs](https://github.com/linq2db/linq2db/blob/master/Tests/Linq/Linq/ValueConversionTests.cs)

Here is the test itself:
```c#
[Table("SymmetryTestTable")]
public class SymmetryTestClass
{
	public DateTime? DateTimeNullable { get; set; }
}

[Test]
public void SymmetryTest([DataSources(false)] string context)
{
	var ms = new MappingSchema();
	ms.GetFluentMappingBuilder()
		.Entity<SymmetryTestClass>()
			.Property(e => e.DateTimeNullable)
				.HasConversion(
					_ => _.HasValue ? _.Value.ToLocalTime() : new DateTime?(),
					_ => _.HasValue ? new DateTime(_.Value.Ticks, DateTimeKind.Local) : new DateTime?()
				);

	using var db = GetDataContext(context, ms);
	using var table = db.CreateLocalTable(Array.Empty<SymmetryTestClass>());

	DateTime now = DateTime.Now;

	Assert.DoesNotThrow(() => table.Where(e => e.DateTimeNullable.Value == now).ToArray());
	Assert.DoesNotThrow(() => table.Where(e => now == e.DateTimeNullable.Value).ToArray());
}
```

Note that **the latter assert does fail**.

### PR DESCRIPTION
**Pull request by sdanyliv**
Fixes #2620

### CODE REVIEW

File: Tests/Linq/Linq/ValueConversionTests.cs [hunk 1]
```diff
@@ -70,7 +73,8 @@ public static MainClass[] TestData()
 							EnumNullable = i % 4 == 1 ? EnumValue.Value1 : i % 4 == 2 ? EnumValue.Value2 : i % 4 == 3 ? EnumValue.Value3 : (EnumValue?)null,
 							EnumWithNull = i % 4 == 1 ? EnumValue.Value1 : i % 4 == 2 ? EnumValue.Value2 : i % 4 == 3 ? EnumValue.Value3 : EnumValue.Null,
 							EnumWithNullDeclarative = i % 4 == 1 ? EnumValue.Value1 : i % 4 == 2 ? EnumValue.Value2 : i % 4 == 3 ? EnumValue.Value3 : EnumValue.Null,
-							BoolValue = i % 4 == 1
+							BoolValue = i % 4 == 1,
+							DateTimeNullable = i % 3 == 1 ? (DateTime?)null : new DateTime(2020, 11, 13)
```
Comment (MaceWindu) [inline Tests/Linq/Linq/ValueConversionTests.cs:77]: better to use existing constant `TestData.Date`

File: Tests/Linq/Linq/ValueConversionTests.cs [hunk 2]
```diff
@@ -246,6 +258,32 @@ select new
...
+		[Test]
+		public void ParameterTestsNullable([IncludeDataSources(false, TestProvName.AllSQLite)] string context)
+		{
+			var ms = CreateMappingSchema();
+
+			var testData = MainClass.TestData();
+			using (var db = GetDataContext(context, ms))
+			using (var table = db.CreateLocalTable(testData))
+			{
+				var testDate = new DateTime(2020, 11, 13);
```
Comment (MaceWindu) [inline Tests/Linq/Linq/ValueConversionTests.cs:270]: same here

**Review by MaceWindu — APPROVED** [review]

**Comment by viceroypenguin** [comment]
Rename PR ("hullable" should be "nullable")?

**Review by viceroypenguin — APPROVED** [review]

**Comment by MaceWindu** [comment]
that's insane test failure. investigating, but it doesn't make any sense...

**Comment by MaceWindu** [comment]
hmm, from other side it could be related to nullability changes, but still it doesn't make  much sense

**Comment by MaceWindu** [comment]
that's really cool. addition of datetime column uncovered two bugs

File: Source/LinqToDB/Data/CommandInfo.cs [hunk 3]
```diff
@@ -1536,7 +1536,7 @@ static string GetFieldsKey(IDataReader dataReader)
 			var key = new QueryKey(typeof(T), dataReader.GetType(), dataConnection.ID, sql, additionalKey);
 
 			var func = CreateObjectReader<T>(dataConnection, dataReader, (dc, dr, type, idx,dataReaderExpr) =>
-				new ConvertFromDataReaderExpression(type, idx, null, dataReaderExpr).Reduce(dc));
+				new ConvertFromDataReaderExpression(type, idx, null, dataReaderExpr).Reduce(dc, true));
```
Comment (viceroypenguin) [inline Source/LinqToDB/Data/CommandInfo.cs:1539]:
```suggestion
				new ConvertFromDataReaderExpression(type, idx, null, dataReaderExpr).Reduce(dc, slowMode: true));
```

File: Source/LinqToDB/DataProvider/ReaderInfo.cs [hunk 4]
```diff
@@ -50,7 +50,7 @@ void CalcHashCode()
 				_hashCode = _hashCode * -1521134295 + (ToType            == null ? 0 : ToType.           GetHashCode());
 				_hashCode = _hashCode * -1521134295 + (FieldType         == null ? 0 : FieldType.        GetHashCode());
 				_hashCode = _hashCode * -1521134295 + (ProviderFieldType == null ? 0 : ProviderFieldType.GetHashCode());
-				_hashCode = _hashCode * -1521134295 + (DataTypeName      == null ? 0 : DataTypeName.     GetHashCode());
+				_hashCode = _hashCode * -1521134295 + (DataTypeName      == null ? 0 : DataTypeName     .GetHashCode());
 				_hashCode = _hashCode * -1521134295 + (DataReaderType    == null ? 0 : DataReaderType.   GetHashCode());
```
Comment (viceroypenguin) [inline Source/LinqToDB/DataProvider/ReaderInfo.cs:54]:
```suggestion
				_hashCode = _hashCode * -1521134295 + (DataReaderType    == null ? 0 : DataReaderType   .GetHashCode());
```

File: Source/LinqToDB/Expressions/ConvertFromDataReaderExpression.cs [hunk 5]
```diff
@@ -45,25 +45,25 @@ class ConvertFromDataReaderExpression : Expression
...
 		public Expression Reduce(IDataContext dataContext, IDataReader dataReader)
 		{
-			return GetColumnReader(dataContext, dataContext.MappingSchema, dataReader, _type, Converter, _idx, _dataReaderParam);
+			return GetColumnReader(dataContext, dataContext.MappingSchema, dataReader, _type, Converter, _idx, _dataReaderParam, false);
 		}
 
 		public Expression Reduce(IDataContext dataContext, IDataReader dataReader, Expression dataReaderParam)
 		{
-			return GetColumnReader(dataContext, dataContext.MappingSchema, dataReader, _type, Converter, _idx, dataReaderParam);
+			return GetColumnReader(dataContext, dataContext.MappingSchema, dataReader, _type, Converter, _idx, dataReaderParam, false);
```
Comment (viceroypenguin) [inline Source/LinqToDB/Expressions/ConvertFromDataReaderExpression.cs:66]:
```suggestion
			return GetColumnReader(dataContext, dataContext.MappingSchema, dataReader, _type, Converter, _idx, dataReaderParam, forceNullCheck: false);
```

File: Source/LinqToDB/Expressions/ConvertFromDataReaderExpression.cs [hunk 6]
```diff
@@ -45,25 +45,25 @@ class ConvertFromDataReaderExpression : Expression
...
+			var columnReader = new ColumnReader(dataContext, dataContext.MappingSchema, _type, _idx, Converter, slowMode);
 			return Convert(Call(Constant(columnReader), _columnReaderGetValueInfo, _dataReaderParam), _type);
 		}
 
 		static readonly MethodInfo _isDBNullInfo = MemberHelper.MethodOf<IDataReader>(rd => rd.IsDBNull(0));
 
 		public Expression Reduce(IDataContext dataContext, IDataReader dataReader)
 		{
-			return GetColumnReader(dataContext, dataContext.MappingSchema, dataReader, _type, Converter, _idx, _dataReaderParam);
+			return GetColumnReader(dataContext, dataContext.MappingSchema, dataReader, _type, Converter, _idx, _dataReaderParam, false);
```
Comment (viceroypenguin) [inline Source/LinqToDB/Expressions/ConvertFromDataReaderExpression.cs:61]:
```suggestion
			return GetColumnReader(dataContext, dataContext.MappingSchema, dataReader, _type, Converter, _idx, _dataReaderParam, forceNullCheck: false);
```

File: Source/LinqToDB/ServiceModel/LinqService.cs [hunk 7]
```diff
@@ -218,7 +218,7 @@ public string ExecuteReader(string? configuration, string queryData)
 
 				for (var i = 0; i < ret.FieldCount; i++)
 					columnReaders[i] = new ConvertFromDataReaderExpression.ColumnReader(db, db.MappingSchema,
-						ret.FieldTypes[i], i, QueryHelper.GetValueConverter(select.Select.Columns[i]));
+						ret.FieldTypes[i], i, QueryHelper.GetValueConverter(select.Select.Columns[i]), true);
```
Comment (viceroypenguin) [inline Source/LinqToDB/ServiceModel/LinqService.cs:221]: parameter name?

**Pull request merged**